src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let stateCopy = { ...state };
+
+  for (const impact of actions) {
+    switch (impact.type) {
+      case 'addProperties':
+        stateCopy = { ...stateCopy };
+        Object.assign(stateCopy, impact['extraData']);
+        break;
+
+      case 'removeProperties':
+        stateCopy = { ...stateCopy };
+
+        for (const key of impact.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      default:
+        break;
+    }
+    result.push(stateCopy);
```

Comment:
you can create a static state right here instead of `stateCopy = { ...stateCopy };` on each case:

```suggestion
    result.push({ ...stateCopy });
```